013-generalize-address-spaces.md

```diff
@@ -2,429 +2,372 @@
+
+The attribute has also been used for subprograms and subroutine types,
+to specify what addressing mode should be used to call the subprogram.
+
+While it is similar to address classes in that a pointer type for one
+address space may be a different size from that for another address
+space, the address space is not part of the pointer value, but is
+implicit in the type of the pointer. The bits of the pointer refer to
+an address in a completely separate address space.
+
+### Why `DW_OP_xderef` is not sufficient
+
+The DWARF 5 `DW_OP_xderef*` operations allow a value to be converted
+into an address within a specified address space which is then
+read. But it provides no way to create a memory location for an
+address in the non-default address space. For example, GPU variables
+can be allocated in local scratch pad memory at a fixed address.
+
+### Address spaces are not just swizzled pointers
+
+While some hardware's memory controllers may implement different
+address spaces by swizzling pointers, it cannot be assumed that this
+is a universal hardware design. Different GPU designs could have a
+completely separate connection to a particular pool of memory and have
+a completely different numbering system to reference it. On the other
+hand there is nothing in this design which precludes hardware from
+using swizzled pointers to implement address spaces.
+
+If address spaces were implemented by a consumer by swizzling
+pointers, then there would likely need to be reserved ranges of
+addresses defined in the ABI which could interfere with other
+addresses in the system memory. That would require the consumer to
+have special treatment for such values. Furthermore, purely swizzled
+pointers do not capture the semantic differences of address spaces
+like context dependence, pointer size, or embedded access patterns.
+
+**Insert basic example of OMP mapping to GPU memory.
+
+## Proposed solution
+
+Add a new attribute `DW_AT_address_space` to pointer and reference
+types.  This allows the compiler to specify which address space is
+being used to represent the pointer or reference type.
+
+**For discussion** This probably needs more explanation. I wrote about
+it in my reply to Cary. I think this basically requires Baris's
+refined type proposal and this attribute should be applied to all GPU
+pointers. I believe that this is consistent with Tony's original
+idea. I think we need an example in Appendix D showing how this should
+work.
+
+Add a new `DW_OP_mem` operation defined to create a memory location
+description from an address and address space. It can be used to
+specify the location of a variable that is allocated in a specific
+address space.
+
+**For discussion** Why isn't this a flavor of addr. We probably need
+to cover relocation here.
+
+**For Discussion** Is the OpenCL reasoning that Tony cited compelling
+enough that the address space can't be an operand. If so we should
+probably state that in the document.
```

Comment:
The example is that OpenCL has primitives that take the address space as a runtime value. So I think there needs to be a way to allow computed address space IDs. But there is the very common case that the address is statically known, and that is where an xaddr-like operation that works on the non-default address space could make sense.